### PR TITLE
Add oatpp and oatpp-test (both in one repo)

### PR DESCRIPTION
**Pull request by Totto16**
Add oatpp and oatpp-test 
they are both in the same [repo](https://github.com/oatpp/oatpp)

**Note**: I plan to also add some modules that `oatpp` has, since oatpp has many modules, so that without them, it's not really useable.

### CODE REVIEW

**Comment by Totto16** [comment]
I wonder why the tests fail here, in the CI in the fork [here](https://github.com/Totto16/wrapdb/actions/runs/5091880316/jobs/9152489647) they don't fail :thinking:

**Comment by jpakkane** [comment]
The relevant line is probably this:

    FINISHED - failed, leakingObjects = 2

**Comment by Totto16** [comment]
> The relevant line is probably this:
> 
> ```
> FINISHED - failed, leakingObjects = 2
> ```

Yeah, that's the reason why they fail, but that implies some wrong code, or some other problem, which never occured in the CI cmake test of the project itself or in any CI in the fork, which seems strangely odd

**Comment by jpakkane** [comment]
Typically these things imply some sort of UB. If possible, try compiling the code with address sanitizer and then running the tests locally. If it reports something, then that is probably the reason (and if not, fixing the reported issues is useful anyway).

**Comment by Totto16** [comment]
> Typically these things imply some sort of UB. If possible, try compiling the code with address sanitizer and then running the tests locally. If it reports something, then that is probably the reason (and if not, fixing the reported issues is useful anyway).

Thanks, its only suspicios, since it works in any upstream tests, but UB comes in some unexcpetced cases :man_shrugging: And I can't test it locally in a n easy way, so Windows VM which sucks all my RAM away, here I come :sweat: :man_facepalming:

**Comment by Totto16** [comment]
After some testing I figured, that nothing is wrong, it all works as expected, with every sanitizer (memory, address) 
So i dug into the code and a issue, that resolved this problem earlier, [here](https://github.com/oatpp/oatpp/pull/72) is the PR fixing that. And  the problem is, that the tests don't wait on some event, but 1 second (hard coded) [here](https://github.com/oatpp/oatpp/blob/016d47044f81b27d5b26c4fd507b25d41504430e/test/oatpp/web/FullTest.cpp#L226). On some slower machine the teardown of some other things in separate threads may be slower, so it detects `leakingObjects`.

So the issues is only, that the tests aren't well written and the CI is sometimes to slow, it's really bad, but maybe just restart the CI, until it works and we are lucky, we could modify the source code with diff files in the .wrap file, but thats really dirty, and for tests it's really ugly...

So what should we do :thinking:

**Comment by Totto16** [comment]
> I have not looked at the whole code but in C++ if you have spawned threads then you either need to daemonize them or call `.join()`. The former is not recommended outside of some exceptinol circumstances and if you do the latter then you don't need sleeps. If memory servers not calling `join` on a thread is UB so if your code does not do that, then you really should fix that.

They deal with that in the code correctly, but since they do some other stuff in the threads, it doesn't always take 1 second to "deallocate" resources, on linux it's fast enough, but threads in windows are slower in general, so the application itself is flawless in that regards. But the tests are hacky, since there's no way to know (as user of the program) when a connection in a separate thread is cleaned up, it is assured, that it is cleaned up, before it does the shutdown. But the tests are hacky and not well written. They just assume, that it happened and assert that. But that only works, when that happens fast enough. They do not shutdown the programm, where a connection cleanup is assured, so that problem occurs. 

So tldr, the application is ok, the tests are poorly written.

Fixing it upstream would surely be possible, but the problem is, that no new release would be made (the release 1.3.0 was released 19.11.2021) even if the application is maintained, no regular releases are made, so we'd have to switch to wrap-git :sweat:

**Comment by jpakkane** [comment]
Can you make it so that the test is only run if the version is less than 1.4? So it would currently not be run but once the next release comes out it errors out if it has not been fixed upstream.

**Comment by Totto16** [comment]
> Can you make it so that the test is only run if the version is less than 1.4? So it would currently not be run but once the next release comes out it errors out if it has not been fixed upstream.

yeah that would be possible :+1:

**Comment by Totto16** [comment]
I now solved it, by using a feature for the tests, so that it can detect, if the tests likely will succeed, and enabled them automatically, but allow the user to turn them off or on explicitly

File: subprojects/packagefiles/oatpp/test/meson.build [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    'oatpp/web/server/handler/AuthorizationHandlerTest.cpp',
+    'oatpp/web/server/handler/AuthorizationHandlerTest.hpp',
+)
+
+
+all_tests = executable(
+    'oatpp_all_tests',
+    test_src_files,
+    cpp_args: flags,
+    include_directories: include_directories('.'),
```
Comment (jpakkane) [inline subprojects/packagefiles/oatpp/test/meson.build:118]: You never need to do this. The current source and build directories are put in the search path by default.

**Comment by jpakkane** [comment]
In the future you should keep your MR squashed so that it only has one commit. I'll squash this for now.

File: subprojects/packagefiles/oatpp/test/meson.build [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+all_tests = executable(
+    'oatpp_all_tests',
+    test_src_files,
+    cpp_args: flags,
+    dependencies: [oatpp_dep, oatpp_test_dep],
+)
+# Note that these are extensive tests and take a long time! It takes ~ 5 Min on fast machines
+test('oatpp all tests', all_tests, timeout: 600)
```
[the remark this answers is not in the document]
Comment (Totto16) [reply subprojects/packagefiles/oatpp/test/meson.build:121]: I would say around 900 seconds, these are extensive tests and In Linux on my machine they run in under 300 seconds. but I have a powerful machine. In my VM for windows they took mostly around 500 secs, but sometimes, when many things happened in the background up to ~800 secs :sweat: 

I figured that meson has an option test-multiplier, but it's not passable as CLI arg, or have I misread something ther?

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply subprojects/packagefiles/oatpp/test/meson.build:121]: It's possible with the PR I just merged that adds test_options to the ci_config.json file.

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply subprojects/packagefiles/oatpp/test/meson.build:121]: Also I wonder how parallel the upstream tests are. Meson can automatically parallelize tests if they are run as separate units, but this test bundles everything in one binary because the main() function runs every single test one after the other. :D Did not check further to see if it has e.g. threading.

[the remark this answers is not in the document]
Comment (Totto16) [reply subprojects/packagefiles/oatpp/test/meson.build:121]: The tests use threading extensively, they spawn many threads and run them in parallel, so single threading is likely to reduce the performance of the tests massively

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply subprojects/packagefiles/oatpp/test/meson.build:121]: Gotcha, so nothing to solve there.

I guess it really is just doing "that much work".
